{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "MODELS = [\n",
    "    \"swiss_model\",\n",
    "    \"colab_alphafold2\",\n",
    "    \"modeller\",\n",
    "    \"phyre2\",\n",
    "    \"i_tasser\",\n",
    "    \"rosetta\",\n",
    "    \"alphafold3\",\n",
    "    \"foldX\"\n",
    "]\n",
    "\n",
    "CSV_PATH = \"../../data/csv/\"  # Define the path to the CSV files\n",
    "PDB_PATH = \"../../data/pdb/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>swiss_model</th>\n",
       "      <th>phyre2</th>\n",
       "      <th>colab_alphafold2</th>\n",
       "      <th>i_tasser</th>\n",
       "      <th>modeller</th>\n",
       "      <th>rosetta</th>\n",
       "      <th>alphafold3</th>\n",
       "      <th>fold_x</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>https://yanglab.qd.sdu.edu.cn/trRosetta/output...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_completed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Asp28Ala</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>https://yanglab.qd.sdu.edu.cn/trRosetta/output...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_completed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Asp28Gly</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGGGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>https://yanglab.qd.sdu.edu.cn/trRosetta/output...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_completed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Asp28Val</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGVGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>https://yanglab.qd.sdu.edu.cn/trRosetta/output...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_completed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Glu61Asp</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>https://yanglab.qd.sdu.edu.cn/trRosetta/output...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_completed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene identifier          variant  \\\n",
       "0  atpE     Rv1305  atpE_p.Ala63Pro   \n",
       "1  atpE     Rv1305  atpE_p.Asp28Ala   \n",
       "2  atpE     Rv1305  atpE_p.Asp28Gly   \n",
       "3  atpE     Rv1305  atpE_p.Asp28Val   \n",
       "4  atpE     Rv1305  atpE_p.Glu61Asp   \n",
       "\n",
       "                                               fasta swiss_model     phyre2  \\\n",
       "0  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "1  MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "2  MDPTIAAGALIGGGLIMAGGAIGAGIGGGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "3  MDPTIAAGALIGGGLIMAGGAIGAGIGVGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "4  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "\n",
       "  colab_alphafold2       i_tasser   modeller  \\\n",
       "0        concluded  not_concluded  concluded   \n",
       "1        concluded  not_concluded  concluded   \n",
       "2        concluded  not_concluded  concluded   \n",
       "3        concluded  not_concluded  concluded   \n",
       "4        concluded  not_concluded  concluded   \n",
       "\n",
       "                                             rosetta alphafold3         fold_x  \n",
       "0  https://yanglab.qd.sdu.edu.cn/trRosetta/output...  concluded  not_completed  \n",
       "1  https://yanglab.qd.sdu.edu.cn/trRosetta/output...  concluded  not_completed  \n",
       "2  https://yanglab.qd.sdu.edu.cn/trRosetta/output...  concluded  not_completed  \n",
       "3  https://yanglab.qd.sdu.edu.cn/trRosetta/output...  concluded  not_completed  \n",
       "4  https://yanglab.qd.sdu.edu.cn/trRosetta/output...  concluded  not_completed  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df = pd.read_csv(f'{CSV_PATH}fasta_variant.csv', sep=';')  # Read the CSV file into a DataFrame\n",
    "variant_df.head()  # Display the first few rows of the DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>model</th>\n",
       "      <th>pdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>swiss_model</td>\n",
       "      <td>../../data/pdb/swiss_model/atpE_Ala63Pro.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>colab_alphafold2</td>\n",
       "      <td>../../data/pdb/colab_alphafold2/atpE_Ala63Pro.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>modeller</td>\n",
       "      <td>../../data/pdb/modeller/atpE_Ala63Pro.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>phyre2</td>\n",
       "      <td>../../data/pdb/phyre2/atpE_Ala63Pro.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>i_tasser</td>\n",
       "      <td>File not found</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene identifier        variant  \\\n",
       "0  atpE     Rv1305  atpE_Ala63Pro   \n",
       "1  atpE     Rv1305  atpE_Ala63Pro   \n",
       "2  atpE     Rv1305  atpE_Ala63Pro   \n",
       "3  atpE     Rv1305  atpE_Ala63Pro   \n",
       "4  atpE     Rv1305  atpE_Ala63Pro   \n",
       "\n",
       "                                               fasta             model  \\\n",
       "0  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...       swiss_model   \n",
       "1  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...  colab_alphafold2   \n",
       "2  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...          modeller   \n",
       "3  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...            phyre2   \n",
       "4  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...          i_tasser   \n",
       "\n",
       "                                                 pdb  \n",
       "0       ../../data/pdb/swiss_model/atpE_Ala63Pro.pdb  \n",
       "1  ../../data/pdb/colab_alphafold2/atpE_Ala63Pro.pdb  \n",
       "2          ../../data/pdb/modeller/atpE_Ala63Pro.pdb  \n",
       "3            ../../data/pdb/phyre2/atpE_Ala63Pro.pdb  \n",
       "4                                     File not found  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = []\n",
    "columns = variant_df.columns\n",
    "\n",
    "for index, row in variant_df.iterrows():\n",
    "    gene = row[\"gene\"]\n",
    "    identifier = row[\"identifier\"]\n",
    "    variant = row[\"variant\"].replace('_p.', '_')\n",
    "    fasta = row[\"fasta\"]\n",
    "    for model in MODELS:\n",
    "        if model in columns:\n",
    "            status_model = row[model]\n",
    "            if status_model == \"concluded\":\n",
    "                pdb = os.path.join(PDB_PATH, model, f\"{variant}.pdb\")  \n",
    "            else:\n",
    "                pdb = \"File not found\"\n",
    "        else:\n",
    "            pdb = \"File not found\"\n",
    "        \n",
    "        data.append([gene, identifier, variant, fasta, model, pdb])\n",
    "\n",
    "validation_df = pd.DataFrame(data, columns=[\"gene\", \"identifier\", \"variant\", \"fasta\", \"model\", \"pdb\"])\n",
    "validation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if already exist a validation.csv\n",
    "if not os.path.exists(f\"{CSV_PATH}validation.csv\"):\n",
    "    validation_df.to_csv(f\"{CSV_PATH}validation.csv\", index=False, sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
