{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.chrome.service import Service\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.common.exceptions import StaleElementReferenceException\n",
    "\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "CSV_PATH = \"../../data/csv/\"  # Define the path to the CSV files\n",
    "PDB_PATH = \"../../data/pdb/\" # Define the path to the pdb files\n",
    "SAVES_PATH = \"../../data/saves/\" # Define the path to the saves files\n",
    "\n",
    "options = webdriver.ChromeOptions() \n",
    "options.add_argument('--headless') \n",
    "options.add_argument('--no-sandbox') \n",
    "options.add_argument('--disable-dev-shm-usage') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_file(driver, row):\n",
    "    model = row['model']\n",
    "    variant = row['variant']\n",
    "\n",
    "    # if variant == \"atpE_Glu61Asp\" and model == \"alphafold3\":\n",
    "    #     input() \n",
    "    pdb_file = row['pdb']\n",
    "    file_path = os.path.abspath(pdb_file)\n",
    "\n",
    "    wait = WebDriverWait(driver, 12600000)\n",
    "    driver.get('https://saves.mbi.ucla.edu/')\n",
    "    file_field = wait.until(EC.presence_of_element_located((By.ID, \"pdbfile\")))\n",
    "    file_field.send_keys(file_path)\n",
    "    \n",
    "    # click buttons\n",
    "    start_button = wait.until(EC.element_to_be_clickable((By.ID, 'startjob')))\n",
    "    driver.execute_script(\"arguments[0].scrollIntoView(true);\", start_button)\n",
    "    start_button.click()\n",
    "    \n",
    "    errat_button = wait.until(EC.element_to_be_clickable((By.ID, 'errat')))\n",
    "    driver.execute_script(\"arguments[0].scrollIntoView(true);\", errat_button)\n",
    "    errat_button.click()\n",
    "    \n",
    "    verify_button = wait.until(EC.element_to_be_clickable((By.ID, 'verify')))\n",
    "    driver.execute_script(\"arguments[0].scrollIntoView(true);\", verify_button)\n",
    "    verify_button.click()\n",
    "\n",
    "    # wait results\n",
    "    xpath = \"//u[text()='Overall Quality Factor']\"\n",
    "    wait.until(EC.presence_of_element_located((By.XPATH, xpath)))\n",
    "    xpath = \"/html/body/table/tbody/tr[1]/td[2]/div[1]/div/span\"\n",
    "    wait.until(EC.presence_of_element_located((By.XPATH, xpath)))\n",
    "\n",
    "    driver.save_screenshot(f'{SAVES_PATH}{model}/{variant}.png')\n",
    "\n",
    "    # get values\n",
    "    xpath = '/html/body/table/tbody/tr[1]/td[1]/div[1]/div/center/center/h1'\n",
    "    errat_value = wait.until(EC.presence_of_element_located((By.XPATH, xpath))).text\n",
    "    \n",
    "    xpath = '/html/body/table/tbody/tr[1]/td[2]/div[1]/div/center/div[1]'\n",
    "    raw_value = wait.until(EC.presence_of_element_located((By.XPATH, xpath))).text\n",
    "    verify_value = f\"{raw_value.split(' ')[0]} {raw_value.split(' ')[-2]} {raw_value.split(' ')[-1]}\"\n",
    "    return errat_value, verify_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>model</th>\n",
       "      <th>pdb</th>\n",
       "      <th>errat</th>\n",
       "      <th>verify</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>swiss_model</td>\n",
       "      <td>../../data/pdb/swiss_model/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>29.63% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>colab_alphafold2</td>\n",
       "      <td>../../data/pdb/colab_alphafold2/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>48.15% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>modeller</td>\n",
       "      <td>../../data/pdb/modeller/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>28.40% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>phyre2</td>\n",
       "      <td>../../data/pdb/phyre2/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>24.69% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>i_tasser</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene identifier        variant  \\\n",
       "0  atpE     Rv1305  atpE_Ala63Pro   \n",
       "1  atpE     Rv1305  atpE_Ala63Pro   \n",
       "2  atpE     Rv1305  atpE_Ala63Pro   \n",
       "3  atpE     Rv1305  atpE_Ala63Pro   \n",
       "4  atpE     Rv1305  atpE_Ala63Pro   \n",
       "\n",
       "                                               fasta             model  \\\n",
       "0  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...       swiss_model   \n",
       "1  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...  colab_alphafold2   \n",
       "2  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...          modeller   \n",
       "3  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...            phyre2   \n",
       "4  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...          i_tasser   \n",
       "\n",
       "                                                 pdb  errat         verify  \n",
       "0       ../../data/pdb/swiss_model/atpE_Ala63Pro.pdb  100.0  29.63% >= 0.1  \n",
       "1  ../../data/pdb/colab_alphafold2/atpE_Ala63Pro.pdb  100.0  48.15% >= 0.1  \n",
       "2          ../../data/pdb/modeller/atpE_Ala63Pro.pdb  100.0  28.40% >= 0.1  \n",
       "3            ../../data/pdb/phyre2/atpE_Ala63Pro.pdb  100.0  24.69% >= 0.1  \n",
       "4                                     File not found    NaN            NaN  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validation_df = pd.read_csv(f'{CSV_PATH}validation.csv', sep=';')  # Read the CSV file into a DataFrame\n",
    "validation_df.head()  # Display the first few rows of the DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3072"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(validation_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total: 3072\n",
      " Done: 1813\n",
      " Todo: 1259\n"
     ]
    }
   ],
   "source": [
    "total = len(validation_df)\n",
    "todo = int(validation_df[\"errat\"].isna().sum())\n",
    "done = total - todo\n",
    "print(f\"Total: {total}\\n Done: {done}\\n Todo: {todo}\")\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>model</th>\n",
       "      <th>pdb</th>\n",
       "      <th>errat</th>\n",
       "      <th>verify</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2182</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>rpoB_Arg448Lys</td>\n",
       "      <td>LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...</td>\n",
       "      <td>alphafold3</td>\n",
       "      <td>../../data/pdb/alphafold3/rpoB_Arg448Lys.pdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2190</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>rpoB_Asn437Asp</td>\n",
       "      <td>LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...</td>\n",
       "      <td>alphafold3</td>\n",
       "      <td>../../data/pdb/alphafold3/rpoB_Asn437Asp.pdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2198</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>rpoB_Asn437His</td>\n",
       "      <td>LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...</td>\n",
       "      <td>alphafold3</td>\n",
       "      <td>../../data/pdb/alphafold3/rpoB_Asn437His.pdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>rpoB_Asn437Ile</td>\n",
       "      <td>LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...</td>\n",
       "      <td>alphafold3</td>\n",
       "      <td>../../data/pdb/alphafold3/rpoB_Asn437Ile.pdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2214</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>rpoB_Asn437Ser</td>\n",
       "      <td>LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...</td>\n",
       "      <td>alphafold3</td>\n",
       "      <td>../../data/pdb/alphafold3/rpoB_Asn437Ser.pdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene identifier         variant  \\\n",
       "2182  rpoB     Rv0667  rpoB_Arg448Lys   \n",
       "2190  rpoB     Rv0667  rpoB_Asn437Asp   \n",
       "2198  rpoB     Rv0667  rpoB_Asn437His   \n",
       "2206  rpoB     Rv0667  rpoB_Asn437Ile   \n",
       "2214  rpoB     Rv0667  rpoB_Asn437Ser   \n",
       "\n",
       "                                                  fasta       model  \\\n",
       "2182  LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...  alphafold3   \n",
       "2190  LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...  alphafold3   \n",
       "2198  LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...  alphafold3   \n",
       "2206  LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...  alphafold3   \n",
       "2214  LADSRQSKTAASPSPSRPQSSSNNSVPGAPNRVSFAKLREPLEVPG...  alphafold3   \n",
       "\n",
       "                                               pdb  errat verify  \n",
       "2182  ../../data/pdb/alphafold3/rpoB_Arg448Lys.pdb    NaN  Error  \n",
       "2190  ../../data/pdb/alphafold3/rpoB_Asn437Asp.pdb    NaN  Error  \n",
       "2198  ../../data/pdb/alphafold3/rpoB_Asn437His.pdb    NaN  Error  \n",
       "2206  ../../data/pdb/alphafold3/rpoB_Asn437Ile.pdb    NaN  Error  \n",
       "2214  ../../data/pdb/alphafold3/rpoB_Asn437Ser.pdb    NaN  Error  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "error_df = validation_df[(validation_df[\"errat\"]==\"Error\") | (validation_df[\"verify\"]==\"Error\")]\n",
    "error_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>model</th>\n",
       "      <th>pdb</th>\n",
       "      <th>errat</th>\n",
       "      <th>verify</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>i_tasser</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>rosetta</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>foldX</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Asp28Ala</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>i_tasser</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Asp28Ala</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>rosetta</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene identifier        variant  \\\n",
       "4   atpE     Rv1305  atpE_Ala63Pro   \n",
       "5   atpE     Rv1305  atpE_Ala63Pro   \n",
       "7   atpE     Rv1305  atpE_Ala63Pro   \n",
       "12  atpE     Rv1305  atpE_Asp28Ala   \n",
       "13  atpE     Rv1305  atpE_Asp28Ala   \n",
       "\n",
       "                                                fasta     model  \\\n",
       "4   MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...  i_tasser   \n",
       "5   MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...   rosetta   \n",
       "7   MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...     foldX   \n",
       "12  MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...  i_tasser   \n",
       "13  MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...   rosetta   \n",
       "\n",
       "               pdb  errat verify  \n",
       "4   File not found    NaN    NaN  \n",
       "5   File not found    NaN    NaN  \n",
       "7   File not found    NaN    NaN  \n",
       "12  File not found    NaN    NaN  \n",
       "13  File not found    NaN    NaN  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_concluded_df = validation_df[(validation_df['errat'].isna()) | (validation_df['verify'].isna())]\n",
    "not_concluded_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_to_keep = [\"alphafold3\", \"colab_alphafold2\", \"swiss_model\", \"modeller\", \"phyre2\"]\n",
    "not_concluded_df = not_concluded_df[not_concluded_df['model'].isin(models_to_keep)]\n",
    "error_df = error_df[error_df['model'].isin(models_to_keep)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: 107\n"
     ]
    }
   ],
   "source": [
    "print(f\"Error: {int(len(error_df))}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ToDo: 107\n"
     ]
    }
   ],
   "source": [
    "print(f\"ToDo: {int(len(not_concluded_df))}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# not_concluded_df = not_concluded_df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating model 1 from 107\n"
     ]
    }
   ],
   "source": [
    "driver = webdriver.Chrome(options=options)\n",
    "# driver = webdriver.Chrome()\n",
    "try:\n",
    "    for i, (index, row) in enumerate(not_concluded_df.iterrows()):\n",
    "        print(f\"Evaluating model {i+1} from {len(not_concluded_df)}\")\n",
    "\n",
    "        try:\n",
    "            errat_value, verify_value = process_file(driver, row)\n",
    "            validation_df.loc[index, 'errat'] = errat_value\n",
    "            validation_df.loc[index, 'verify'] = verify_value\n",
    "            print(f\"Model {row['model']} {row['variant']} processed\")\n",
    "        except StaleElementReferenceException as e:\n",
    "            print(f\"Stale element reference error processing model {row['model']} {row['variant']}: {e}\")\n",
    "            # Retry the operation\n",
    "            try:\n",
    "                print(f\"Model {row['model']} {row['variant']} retrying\")\n",
    "                errat_value, verify_value = process_file(driver, row)\n",
    "                validation_df.loc[index, 'errat'] = errat_value\n",
    "                validation_df.loc[index, 'verify'] = verify_value\n",
    "            except Exception as e:\n",
    "                print(f\"Error processing model {row['model']} {row['variant']}: {e}\")\n",
    "                validation_df.loc[index, 'errat'] = \"Error\"\n",
    "                validation_df.loc[index, 'verify'] = \"Error\"\n",
    "                # input(\"ERROR\")\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing model {row['model']} {row['variant']}: {e}\")\n",
    "            validation_df.loc[index, 'errat'] = \"Error\"\n",
    "            validation_df.loc[index, 'verify'] = \"Error\"\n",
    "            # input(\"ERROR\")\n",
    "        finally:\n",
    "            if validation_df.loc[index, 'errat']==\"ERROR\" or validation_df.loc[index, 'verify']==\"ERROR\" or validation_df.loc[index, 'errat']==None or validation_df.loc[index, 'verify']==None:\n",
    "                validation_df.loc[index, 'errat'] = \"Error\"\n",
    "                validation_df.loc[index, 'verify'] = \"Error\"\n",
    "            validation_df.to_csv(f\"{CSV_PATH}validation.csv\", index=False, sep=';')\n",
    "finally:\n",
    "    driver.quit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>model</th>\n",
       "      <th>pdb</th>\n",
       "      <th>errat</th>\n",
       "      <th>verify</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>swiss_model</td>\n",
       "      <td>../../data/pdb/swiss_model/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>29.63% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>colab_alphafold2</td>\n",
       "      <td>../../data/pdb/colab_alphafold2/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>48.15% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>modeller</td>\n",
       "      <td>../../data/pdb/modeller/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>28.40% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>phyre2</td>\n",
       "      <td>../../data/pdb/phyre2/atpE_Ala63Pro.pdb</td>\n",
       "      <td>100.0</td>\n",
       "      <td>24.69% &gt;= 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>i_tasser</td>\n",
       "      <td>File not found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene identifier        variant  \\\n",
       "0  atpE     Rv1305  atpE_Ala63Pro   \n",
       "1  atpE     Rv1305  atpE_Ala63Pro   \n",
       "2  atpE     Rv1305  atpE_Ala63Pro   \n",
       "3  atpE     Rv1305  atpE_Ala63Pro   \n",
       "4  atpE     Rv1305  atpE_Ala63Pro   \n",
       "\n",
       "                                               fasta             model  \\\n",
       "0  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...       swiss_model   \n",
       "1  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...  colab_alphafold2   \n",
       "2  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...          modeller   \n",
       "3  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...            phyre2   \n",
       "4  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...          i_tasser   \n",
       "\n",
       "                                                 pdb  errat         verify  \n",
       "0       ../../data/pdb/swiss_model/atpE_Ala63Pro.pdb  100.0  29.63% >= 0.1  \n",
       "1  ../../data/pdb/colab_alphafold2/atpE_Ala63Pro.pdb  100.0  48.15% >= 0.1  \n",
       "2          ../../data/pdb/modeller/atpE_Ala63Pro.pdb  100.0  28.40% >= 0.1  \n",
       "3            ../../data/pdb/phyre2/atpE_Ala63Pro.pdb  100.0  24.69% >= 0.1  \n",
       "4                                     File not found    NaN            NaN  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_df.to_csv(f\"{CSV_PATH}validation.csv\", index=False, sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
