{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Format .ALI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### In order to run Modeller script, you need to change the kernel to not use the virtual env once the Modeller is installed on your machine, click on the kernel on top right of this cell and chose a python kernel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries and loading dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 12 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd  # For data manipulation and analysis\n",
    "from modeller import *  # For working with Modeller tools\n",
    "from modeller.automodel import *  # For automatic model building using Modeller\n",
    "from pandarallel import pandarallel  # For parallel processing of pandas DataFrames\n",
    "import os  # For file and directory operations\n",
    "\n",
    "pandarallel.initialize() # Initializing pandarallel for parallel processing\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define global variables and Modeller environment\n",
    "env = Environ()  # Create a Modeller environment object\n",
    "\n",
    "# File paths for input and output data\n",
    "CSV_PATH = \"../../data/csv/\"  # Path to CSV files\n",
    "PDB_PATH = '../../data/pdb/blast/'  # Path to PDB files\n",
    "PIR_PATH = '../../data/pir/'  # Path to PIR files\n",
    "ALI_PATH = '../../data/ali/'  # Path to ALI files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating .ALI files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>identifier</th>\n",
       "      <th>variant</th>\n",
       "      <th>fasta</th>\n",
       "      <th>swiss_model</th>\n",
       "      <th>phyre2</th>\n",
       "      <th>colab_alphafold2</th>\n",
       "      <th>i_tasser</th>\n",
       "      <th>modeller</th>\n",
       "      <th>rosetta</th>\n",
       "      <th>alphafold3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Ala63Pro</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Asp28Ala</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Asp28Gly</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGGGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Asp28Val</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGVGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atpE</td>\n",
       "      <td>Rv1305</td>\n",
       "      <td>atpE_p.Glu61Asp</td>\n",
       "      <td>MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>not_concluded</td>\n",
       "      <td>concluded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene identifier          variant  \\\n",
       "0  atpE     Rv1305  atpE_p.Ala63Pro   \n",
       "1  atpE     Rv1305  atpE_p.Asp28Ala   \n",
       "2  atpE     Rv1305  atpE_p.Asp28Gly   \n",
       "3  atpE     Rv1305  atpE_p.Asp28Val   \n",
       "4  atpE     Rv1305  atpE_p.Glu61Asp   \n",
       "\n",
       "                                               fasta swiss_model     phyre2  \\\n",
       "0  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "1  MDPTIAAGALIGGGLIMAGGAIGAGIGAGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "2  MDPTIAAGALIGGGLIMAGGAIGAGIGGGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "3  MDPTIAAGALIGGGLIMAGGAIGAGIGVGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "4  MDPTIAAGALIGGGLIMAGGAIGAGIGDGVAGNALISGVARQPEAQ...   concluded  concluded   \n",
       "\n",
       "  colab_alphafold2       i_tasser       modeller        rosetta alphafold3  \n",
       "0        concluded  not_concluded  not_concluded  not_concluded  concluded  \n",
       "1        concluded  not_concluded  not_concluded  not_concluded  concluded  \n",
       "2        concluded  not_concluded  not_concluded  not_concluded  concluded  \n",
       "3        concluded  not_concluded  not_concluded  not_concluded  concluded  \n",
       "4        concluded  not_concluded  not_concluded  not_concluded  concluded  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the variant data from a CSV file\n",
    "variant_df = pd.read_csv(f'{CSV_PATH}fasta_variant.csv', sep=';')  # Load data into a pandas DataFrame\n",
    "variant_df.head()  # Display the first few rows of the DataFrame for verification\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to process each variant and generate .ali files\n",
    "def process_variant(row):\n",
    "    gene = row['gene']  # Extract the gene name from the current row\n",
    "    variant_raw = row['variant']  # Extract the raw variant name from the current row\n",
    "    variant = row['variant'].replace(\"_p.\", \"_\")  # Replace \"_p.\" in the variant name with \"_\"\n",
    "    ali_file_path = os.path.join(ALI_PATH, f\"{variant}.ali\")  # Construct the .ali file path\n",
    "\n",
    "    # Check if the .ali file already exists\n",
    "    if not os.path.exists(ali_file_path):\n",
    "        aln = Alignment(env)  # Create a new alignment object for this variant\n",
    "\n",
    "        # Add the model to the alignment object using the PDB file\n",
    "        md1 = Model(env, file=f'{PDB_PATH}{gene}.pdb')\n",
    "        aln.append_model(md1, align_codes=gene, atom_files=f'{PDB_PATH}{gene}.pdb')\n",
    "\n",
    "        # Add the PIR file to the alignment object\n",
    "        aln.append(file=f'{PIR_PATH}{variant}.txt', align_codes=variant_raw)\n",
    "\n",
    "        # Perform the alignment\n",
    "        aln.align2d()\n",
    "\n",
    "        # Write the alignment to a .ali file in PIR format\n",
    "        aln.write(file=ali_file_path, alignment_format='PIR')\n",
    "\n",
    "        print(f\"Generated alignment file for {variant}\")  # Notify the user of the generated file\n",
    "    else:\n",
    "        # Notify the user if the file already exists\n",
    "        print(f\"Alignment file for {variant} already exists. Skipping.\")\n",
    "\n",
    "    return row  # Return the processed row\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences in CSV: 384\n",
      "Number of .ali files: 384\n",
      "Number of missing files: 0\n"
     ]
    }
   ],
   "source": [
    "# Display the total number of sequences and existing .ali files\n",
    "print(f\"Number of sequences in CSV: {len(variant_df)}\")  # Print the total number of sequences in the CSV\n",
    "print(f\"Number of .ali files: {len([name for name in os.listdir(ALI_PATH) if name.endswith('.ali')])}\")  # Count and print the number of .ali files in the directory\n",
    "print(f\"Number of missing files: {len(variant_df) - len([name for name in os.listdir(ALI_PATH)])}\")  # Calculate and print the number of missing files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment file for pncA_His43Pro already exists. Skipping.Alignment file for atpE_Ala63Pro already exists. Skipping.Alignment file for pncA_Pro54Arg already exists. Skipping.Alignment file for rpoB_Ser450Cys already exists. Skipping.Alignment file for rpoB_Gln429His already exists. Skipping.Alignment file for pncA_Leu172Pro already exists. Skipping.Alignment file for pncA_Val180Ala already exists. Skipping.Alignment file for pncA_Asp8Gly already exists. Skipping.Alignment file for ethA_Asn379Asp already exists. Skipping.Alignment file for pncA_Ala102Pro already exists. Skipping.Alignment file for pncA_Thr76Ile already exists. Skipping.Alignment file for rpoB_Leu449Met already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Pro54Gln already exists. Skipping.Alignment file for atpE_Asp28Ala already exists. Skipping.Alignment file for pncA_Leu182Ser already exists. Skipping.Alignment file for rpoB_Ser450Gln already exists. Skipping.Alignment file for pncA_His51Arg already exists. Skipping.Alignment file for pncA_Asp8His already exists. Skipping.Alignment file for rpoB_Gln429Leu already exists. Skipping.Alignment file for ethA_Cys403Tyr already exists. Skipping.Alignment file for pncA_Val180Gly already exists. Skipping.\n",
      "Alignment file for pncA_Ala102Thr already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Thr76Pro already exists. Skipping.\n",
      "Alignment file for rpoB_Leu452Met already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for atpE_Asp28Gly already exists. Skipping.Alignment file for pncA_His51Asp already exists. Skipping.Alignment file for pncA_Leu182Trp already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Pro54Leu already exists. Skipping.Alignment file for pncA_Asp8Tyr already exists. Skipping.Alignment file for rpoB_Ser450Gly already exists. Skipping.Alignment file for pncA_Ala134Val already exists. Skipping.Alignment file for rpoB_Gln429Pro already exists. Skipping.Alignment file for ethA_Leu35Arg already exists. Skipping.Alignment file for pncA_Val180Phe already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Trp119Arg already exists. Skipping.\n",
      "Alignment file for rpoB_Leu452Pro already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for atpE_Asp28Val already exists. Skipping.Alignment file for pncA_His51Gln already exists. Skipping.\n",
      "Alignment file for pncA_Leu19Pro already exists. Skipping.\n",
      "Alignment file for pncA_Pro54Ser already exists. Skipping.Alignment file for pncA_Cys138Arg already exists. Skipping.Alignment file for pncA_Ala143Gly already exists. Skipping.Alignment file for rpoB_Ser450Leu already exists. Skipping.Alignment file for rpoB_Gln432Asn already exists. Skipping.Alignment file for pncA_Val21Gly already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Trp119Cys already exists. Skipping.Alignment file for ethA_Ser390Phe already exists. Skipping.Alignment file for rpoB_Leu452Val already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for atpE_Glu61Asp already exists. Skipping.Alignment file for pncA_His51Pro already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Leu27Pro already exists. Skipping.\n",
      "Alignment file for pncA_Pro54Thr already exists. Skipping.Alignment file for pncA_Cys14Arg already exists. Skipping.Alignment file for pncA_Val44Gly already exists. Skipping.Alignment file for pncA_Ala146Glu already exists. Skipping.Alignment file for rpoB_Gln432Glu already exists. Skipping.\n",
      "Alignment file for rpoB_Ser450Met already exists. Skipping.\n",
      "Alignment file for rpoB_Lys446Arg already exists. Skipping.Alignment file for pncA_Trp119Gly already exists. Skipping.Alignment file for ethA_Ser57Tyr already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_His51Tyr already exists. Skipping.Alignment file for atpE_Ile66Met already exists. Skipping.Alignment file for pncA_Ala146Thr already exists. Skipping.Alignment file for pncA_Leu35Pro already exists. Skipping.Alignment file for pncA_Cys14Trp already exists. Skipping.Alignment file for pncA_Pro62Arg already exists. Skipping.Alignment file for pncA_Val7Ala already exists. Skipping.Alignment file for rpoB_Gln432His already exists. Skipping.Alignment file for rpoB_Ser450Phe already exists. Skipping.Alignment file for rpoB_Lys446Gln already exists. Skipping.Alignment file for pncA_Trp119Leu already exists. Skipping.Alignment file for ethA_Thr88Ile already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_His57Arg already exists. Skipping.Alignment file for Rv0678_Ala36Val already exists. Skipping.Alignment file for pncA_Pro62Leu already exists. Skipping.Alignment file for pncA_Leu4Ser already exists. Skipping.Alignment file for pncA_Cys72Arg already exists. Skipping.Alignment file for pncA_Ala146Val already exists. Skipping.Alignment file for pncA_Val7Gly already exists. Skipping.Alignment file for rpoB_Lys446Glu already exists. Skipping.Alignment file for rpoB_Gln432Leu already exists. Skipping.Alignment file for rpoB_Ser450Trp already exists. Skipping.Alignment file for ethA_Val202Gly already exists. Skipping.\n",
      "Alignment file for pncA_Trp68Arg already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_His57Asp already exists. Skipping.Alignment file for pncA_Leu4Trp already exists. Skipping.Alignment file for Rv0678_Asn70Asp already exists. Skipping.Alignment file for pncA_Pro62Ser already exists. Skipping.Alignment file for pncA_Cys72Tyr already exists. Skipping.Alignment file for pncA_Val7Leu already exists. Skipping.Alignment file for pncA_Ala171Glu already exists. Skipping.Alignment file for rpoB_Gln432Lys already exists. Skipping.Alignment file for rpoB_Lys446Thr already exists. Skipping.Alignment file for inhA_Ser94Ala already exists. Skipping.Alignment file for pncA_Trp68Cys already exists. Skipping.Alignment file for rpoB_Ser450Tyr already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Leu85Arg already exists. Skipping.Alignment file for pncA_His57Gln already exists. Skipping.Alignment file for pncA_Pro62Thr already exists. Skipping.Alignment file for Rv0678_Cys46Arg already exists. Skipping.Alignment file for pncA_Val7Phe already exists. Skipping.Alignment file for rpoB_Gln432Pro already exists. Skipping.Alignment file for pncA_Ala171Pro already exists. Skipping.Alignment file for pncA_Gln10Arg already exists. Skipping.Alignment file for katG_Ser315Arg already exists. Skipping.Alignment file for rpoB_Met434Arg already exists. Skipping.Alignment file for rpoB_Ser450Val already exists. Skipping.Alignment file for pncA_Trp68Gly already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_His57Leu already exists. Skipping.Alignment file for pncA_Leu85Pro already exists. Skipping.Alignment file for pncA_Pro69Leu already exists. Skipping.Alignment file for rpoB_Gln436Arg already exists. Skipping.Alignment file for pncA_Ala3Glu already exists. Skipping.Alignment file for pncA_Val93Ala already exists. Skipping.Alignment file for Rv0678_Gly121Arg already exists. Skipping.Alignment file for pncA_Gln10His already exists. Skipping.Alignment file for katG_Ser315Asn already exists. Skipping.Alignment file for rpoB_Met434Ile already exists. Skipping.Alignment file for pncA_Trp68Leu already exists. Skipping.Alignment file for rpoB_Thr427Ala already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_His57Tyr already exists. Skipping.Alignment file for pncA_Ser104Arg already exists. Skipping.Alignment file for rpoB_Gln436Asn already exists. Skipping.Alignment file for pncA_Lys48Glu already exists. Skipping.Alignment file for pncA_Val93Leu already exists. Skipping.Alignment file for pncA_Gln10Pro already exists. Skipping.Alignment file for Rv0678_Ile67Ser already exists. Skipping.Alignment file for pncA_Ala46Glu already exists. Skipping.Alignment file for rpoB_Met434Leu already exists. Skipping.Alignment file for katG_Ser315Ile already exists. Skipping.Alignment file for rpoB_Thr427Asn already exists. Skipping.\n",
      "Alignment file for pncA_Tyr103Asp already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_His71Arg already exists. Skipping.\n",
      "Alignment file for pncA_Lys48Thr already exists. Skipping.Alignment file for pncA_Ser104Gly already exists. Skipping.\n",
      "Alignment file for pncA_Val9Gly already exists. Skipping.Alignment file for pncA_Gln141Pro already exists. Skipping.Alignment file for rpoB_Gln436Pro already exists. Skipping.Alignment file for Rv0678_Leu117Arg already exists. Skipping.Alignment file for pncA_Ala46Val already exists. Skipping.Alignment file for rpoB_Met434Thr already exists. Skipping.\n",
      "Alignment file for rpoB_Thr427Gly already exists. Skipping.\n",
      "Alignment file for katG_Ser315Thr already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Tyr103Cys already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Ser104Ile already exists. Skipping.Alignment file for pncA_His71Pro already exists. Skipping.\n",
      "Alignment file for rpoB_Ala451Gly already exists. Skipping.Alignment file for pncA_Lys96Arg already exists. Skipping.Alignment file for pncA_Gly105Asp already exists. Skipping.\n",
      "Alignment file for Rv0678_Leu32Ser already exists. Skipping.Alignment file for rpoB_Gly426Ser already exists. Skipping.Alignment file for pncA_Arg121Pro already exists. Skipping.\n",
      "Alignment file for rpoB_Thr427Ile already exists. Skipping.Alignment file for rpoB_Met434Val already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for katG_Trp328Leu already exists. Skipping.Alignment file for pncA_Tyr103His already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Ser164Pro already exists. Skipping.\n",
      "Alignment file for pncA_His71Tyr already exists. Skipping.\n",
      "Alignment file for rpoB_Ala451Val already exists. Skipping.Alignment file for pncA_Lys96Asn already exists. Skipping.Alignment file for pncA_Gly105Val already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Arg140Pro already exists. Skipping.Alignment file for Rv0678_Met146Thr already exists. Skipping.Alignment file for rpoB_Gly442Glu already exists. Skipping.Alignment file for rpoB_Pro439Ala already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Thr427Pro already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Tyr34Asp already exists. Skipping.Alignment file for gyrA_Ala90Val already exists. Skipping.Alignment file for pncA_Ser59Pro already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Arg448Gln already exists. Skipping.\n",
      "Alignment file for pncA_His82Arg already exists. Skipping.Alignment file for pncA_Arg154Gly already exists. Skipping.Alignment file for pncA_Gly108Arg already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Lys96Gln already exists. Skipping.\n",
      "Alignment file for tlyA_Asn236Lys already exists. Skipping.\n",
      "Alignment file for rpoB_Thr427Ser already exists. Skipping.Alignment file for rpoB_Pro439Leu already exists. Skipping.Alignment file for rpoB_His445Arg already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for gyrA_Asp89Asn already exists. Skipping.\n",
      "Alignment file for pncA_Tyr34Ser already exists. Skipping.Alignment file for pncA_Ser66Pro already exists. Skipping.Alignment file for rpoB_Arg448Leu already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Gly108Glu already exists. Skipping.Alignment file for pncA_Ile133Thr already exists. Skipping.Alignment file for pncA_Arg154Met already exists. Skipping.Alignment file for pncA_Lys96Glu already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for rpoB_His445Asn already exists. Skipping.\n",
      "Alignment file for rpoB_Pro439Ser already exists. Skipping.Alignment file for rpoB_Thr444Ile already exists. Skipping.\n",
      "\n",
      "Alignment file for tlyA_Gly232Asp already exists. Skipping.Alignment file for gyrA_Asp94Ala already exists. Skipping.\n",
      "Alignment file for pncA_Ser67Pro already exists. Skipping.Alignment file for pncA_Tyr64Asp already exists. Skipping.Alignment file for rpoB_Arg448Lys already exists. Skipping.\n",
      "Alignment file for pncA_Gly132Ala already exists. Skipping.\n",
      "Alignment file for pncA_Asp129Asn already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Ile31Ser already exists. Skipping.\n",
      "Alignment file for pncA_Lys96Thr already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for rpoB_Ser428Arg already exists. Skipping.Alignment file for rpoB_Thr444Pro already exists. Skipping.Alignment file for rpoB_His445Asp already exists. Skipping.\n",
      "Alignment file for ddn_Leu49Pro already exists. Skipping.\n",
      "Alignment file for pncA_Tyr95Asp already exists. Skipping.Alignment file for gyrA_Asp94Asn already exists. Skipping.Alignment file for pncA_Gly132Asp already exists. Skipping.Alignment file for pncA_Thr100Pro already exists. Skipping.Alignment file for rpoB_Asn437Asp already exists. Skipping.\n",
      "Alignment file for pncA_Asp12Ala already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Ile31Thr already exists. Skipping.\n",
      "Alignment file for pncA_Met175Arg already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for rpoB_His445Cys already exists. Skipping.Alignment file for rpoB_Ser428Gly already exists. Skipping.\n",
      "Alignment file for embB_Asp328Tyr already exists. Skipping.Alignment file for rpoB_Thr444Ser already exists. Skipping.Alignment file for pncA_Val125Gly already exists. Skipping.Alignment file for pncA_Gly132Cys already exists. Skipping.Alignment file for gyrA_Asp94Gly already exists. Skipping.\n",
      "Alignment file for rpoB_Asn437His already exists. Skipping.Alignment file for pncA_Thr114Pro already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Asp12Asn already exists. Skipping.\n",
      "Alignment file for pncA_Ile5Ser already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Met175Ile already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for rpoB_His445Gln already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Ser428Ile already exists. Skipping.Alignment file for pncA_Val125Phe already exists. Skipping.Alignment file for embB_Asp354Ala already exists. Skipping.Alignment file for rpoB_Val170Phe already exists. Skipping.Alignment file for rpoB_Asn437Ile already exists. Skipping.Alignment file for pncA_Thr135Asn already exists. Skipping.Alignment file for pncA_Gly132Ser already exists. Skipping.Alignment file for pncA_Asp12Glu already exists. Skipping.\n",
      "\n",
      "Alignment file for gyrA_Asp94His already exists. Skipping.\n",
      "Alignment file for pncA_Ile6Thr already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Met175Thr already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_His445Gly already exists. Skipping.Alignment file for rpoB_Ser428Thr already exists. Skipping.Alignment file for gid_Ala134Glu already exists. Skipping.\n",
      "Alignment file for pncA_Val128Gly already exists. Skipping.Alignment file for rpoB_Asn437Ser already exists. Skipping.Alignment file for embB_Gln497Arg already exists. Skipping.Alignment file for pncA_Thr135Pro already exists. Skipping.\n",
      "Alignment file for pncA_Asp12Gly already exists. Skipping.Alignment file for pncA_Gly162Asp already exists. Skipping.\n",
      "\n",
      "Alignment file for gyrA_Asp94Tyr already exists. Skipping.\n",
      "Alignment file for pncA_Ile90Asn already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Met175Val already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_His445Leu already exists. Skipping.Alignment file for rpoB_Ser431Arg already exists. Skipping.\n",
      "\n",
      "Alignment file for gid_Ala200Glu already exists. Skipping.Alignment file for pncA_Val130Gly already exists. Skipping.\n",
      "Alignment file for embB_Gln497Lys already exists. Skipping.Alignment file for pncA_Thr142Ala already exists. Skipping.Alignment file for pncA_Gly17Asp already exists. Skipping.Alignment file for rpoB_Asn437Tyr already exists. Skipping.Alignment file for pncA_Asp136Gly already exists. Skipping.\n",
      "Alignment file for gyrA_Gly88Ala already exists. Skipping.\n",
      "Alignment file for pncA_Ile90Ser already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_His445Phe already exists. Skipping.\n",
      "Alignment file for pncA_Phe106Leu already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for rpoB_Ser431Asn already exists. Skipping.Alignment file for pncA_Val130Met already exists. Skipping.Alignment file for pncA_Thr142Lys already exists. Skipping.Alignment file for gid_Asp67His already exists. Skipping.\n",
      "Alignment file for embB_Gly406Ala already exists. Skipping.\n",
      "Alignment file for rpoB_Asn438His already exists. Skipping.Alignment file for pncA_Gly24Asp already exists. Skipping.Alignment file for pncA_Asp136Val already exists. Skipping.\n",
      "Alignment file for gyrA_Gly88Cys already exists. Skipping.Alignment file for pncA_Ile90Thr already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Phe106Ser already exists. Skipping.Alignment file for rpoB_His445Pro already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for rpoB_Ser431Gly already exists. Skipping.Alignment file for pncA_Val131Gly already exists. Skipping.\n",
      "Alignment file for pncA_Thr142Met already exists. Skipping.Alignment file for gid_Gly73Ala already exists. Skipping.Alignment file for rpoB_Asp435Ala already exists. Skipping.Alignment file for embB_Gly406Asp already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Gly78Asp already exists. Skipping.Alignment file for pncA_Asp49Ala already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for gyrA_Ser91Pro already exists. Skipping.Alignment file for pncA_Leu116Arg already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Phe13Ile already exists. Skipping.\n",
      "Alignment file for rpoB_His445Ser already exists. Skipping.Alignment file for rpoB_Ser431Thr already exists. Skipping.Alignment file for pncA_Val131Phe already exists. Skipping.\n",
      "Alignment file for pncA_Thr142Pro already exists. Skipping.\n",
      "Alignment file for gid_Gly73Glu already exists. Skipping.Alignment file for rpoB_Asp435Asn already exists. Skipping.\n",
      "Alignment file for pncA_Gly78Ser already exists. Skipping.Alignment file for pncA_Asp49Asn already exists. Skipping.\n",
      "\n",
      "Alignment file for embB_Gly406Cys already exists. Skipping.\n",
      "Alignment file for gyrB_Ala504Val already exists. Skipping.Alignment file for pncA_Leu116Pro already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Phe13Leu already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Ser441Ala already exists. Skipping.Alignment file for rpoB_His445Thr already exists. Skipping.Alignment file for pncA_Val139Ala already exists. Skipping.\n",
      "Alignment file for pncA_Thr160Lys already exists. Skipping.\n",
      "Alignment file for rpoB_Asp435Glu already exists. Skipping.Alignment file for gid_His48Gln already exists. Skipping.\n",
      "Alignment file for pncA_Gly78Val already exists. Skipping.Alignment file for pncA_Asp49Glu already exists. Skipping.\n",
      "\n",
      "Alignment file for embB_Gly406Ser already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Leu120Arg already exists. Skipping.Alignment file for gyrB_Asn499Asp already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Phe13Ser already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Ser441Gln already exists. Skipping.Alignment file for rpoB_His445Tyr already exists. Skipping.\n",
      "Alignment file for pncA_Val139Gly already exists. Skipping.\n",
      "Alignment file for pncA_Thr160Pro already exists. Skipping.\n",
      "Alignment file for rpoB_Asp435Gly already exists. Skipping.Alignment file for gid_Pro84Leu already exists. Skipping.\n",
      "Alignment file for pncA_Gly97Arg already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Asp49Gly already exists. Skipping.Alignment file for embB_Met306Ile already exists. Skipping.\n",
      "\n",
      "Alignment file for gyrB_Asn499Thr already exists. Skipping.Alignment file for pncA_Leu120Gln already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Phe58Cys already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Ile491Phe already exists. Skipping.Alignment file for rpoB_Ser441Leu already exists. Skipping.\n",
      "Alignment file for pncA_Thr168Pro already exists. Skipping.Alignment file for pncA_Val139Leu already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Asp63Ala already exists. Skipping.Alignment file for rpoB_Asp435His already exists. Skipping.Alignment file for pncA_Gly97Asp already exists. Skipping.Alignment file for gid_Ser70Asn already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for embB_Met306Leu already exists. Skipping.\n",
      "Alignment file for pncA_Leu120Pro already exists. Skipping.Alignment file for pncA_Phe58Leu already exists. Skipping.\n",
      "Alignment file for gyrB_Asp461Asn already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Leu430Arg already exists. Skipping.Alignment file for pncA_Thr177Pro already exists. Skipping.Alignment file for pncA_Val139Met already exists. Skipping.\n",
      "Alignment file for rpoB_Ser441Lys already exists. Skipping.\n",
      "Alignment file for pncA_Asp63Gly already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Gly97Cys already exists. Skipping.Alignment file for rpoB_Asp435Leu already exists. Skipping.Alignment file for gid_Val105Glu already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Phe58Ser already exists. Skipping.Alignment file for embB_Met306Val already exists. Skipping.Alignment file for pncA_Leu151Ser already exists. Skipping.Alignment file for gyrB_Asp461His already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Val155Ala already exists. Skipping.Alignment file for pncA_Thr47Ala already exists. Skipping.Alignment file for rpoB_Leu430Pro already exists. Skipping.Alignment file for rpoB_Ser441Met already exists. Skipping.Alignment file for pncA_Asp8Ala already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Gly97Ser already exists. Skipping.Alignment file for rpoB_Asp435Phe already exists. Skipping.\n",
      "Alignment file for rpsL_Lys43Arg already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for pncA_Phe94Cys already exists. Skipping.Alignment file for pncA_Leu159Arg already exists. Skipping.\n",
      "Alignment file for embB_Tyr319Cys already exists. Skipping.\n",
      "\n",
      "\n",
      "Alignment file for gyrB_Glu501Asp already exists. Skipping.Alignment file for pncA_Thr47Pro already exists. Skipping.Alignment file for pncA_Val155Gly already exists. Skipping.Alignment file for rpoB_Ser441Trp already exists. Skipping.Alignment file for rpoB_Leu443Phe already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Asp8Asn already exists. Skipping.Alignment file for rpoB_Asp435Tyr already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Gly97Val already exists. Skipping.\n",
      "\n",
      "Alignment file for rpsL_Lys88Arg already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Phe94Leu already exists. Skipping.\n",
      "\n",
      "Alignment file for pncA_Leu159Pro already exists. Skipping.\n",
      "Alignment file for embB_Tyr319Ser already exists. Skipping.Alignment file for gyrB_Glu501Val already exists. Skipping.\n",
      "Alignment file for pncA_Val155Met already exists. Skipping.Alignment file for pncA_Thr61Pro already exists. Skipping.Alignment file for rpoB_Ser441Val already exists. Skipping.\n",
      "Alignment file for rpoB_Leu443Ser already exists. Skipping.Alignment file for pncA_Asp8Glu already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Asp435Val already exists. Skipping.\n",
      "Alignment file for pncA_His137Pro already exists. Skipping.Alignment file for rpsL_Lys88Gln already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for pncA_Phe94Ser already exists. Skipping.\n",
      "Alignment file for pncA_Leu159Val already exists. Skipping.\n",
      "Alignment file for ethA_Ala341Val already exists. Skipping.Alignment file for gyrB_Ser447Phe already exists. Skipping.\n",
      "\n",
      "Alignment file for rpoB_Ser450Ala already exists. Skipping.Alignment file for rpoB_Leu443Trp already exists. Skipping.Alignment file for pncA_Val163Ala already exists. Skipping.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Alignment file for rpsL_Lys88Met already exists. Skipping.\n",
      "Alignment file for ethA_Arg207Gly already exists. Skipping.Alignment file for rplC_Cys154Arg already exists. Skipping.\n",
      "\n",
      "\n",
      "Number of sequences in CSV: 384\n",
      "Number of .ali files: 384\n"
     ]
    }
   ],
   "source": [
    "# Apply the process_variant function to each row in the DataFrame using parallel processing\n",
    "variant_df.parallel_apply(process_variant, axis=1)\n",
    "\n",
    "# Display the updated number of sequences and .ali files after processing\n",
    "print(f\"Number of sequences in CSV: {len(variant_df)}\")  # Print the total number of sequences in the CSV again\n",
    "print(f\"Number of .ali files: {len([name for name in os.listdir(ALI_PATH) if name.endswith('.ali')])}\")  # Count and print the number of .ali files after processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
